lib/srv/heartbeatv2.go

```diff
@@ -499,3 +524,70 @@ func (h *sshServerHeartbeatV2) Announce(ctx context.Context, sender inventory.Do
 	h.prev = server
 	return true
 }
+
+// appServerHeartbeatV2 is the heartbeatV2 implementation for app servers.
+type appServerHeartbeatV2 struct {
+	getServer func(ctx context.Context) *types.AppServerV3
+	announcer auth.Announcer
+	prev      *types.AppServerV3
+}
+
+func (h *appServerHeartbeatV2) Poll(ctx context.Context) (changed bool) {
+	if h.prev == nil {
+		return true
+	}
+	return services.CompareServers(h.getServer(ctx), h.prev) == services.Different
+}
+
+func (h *appServerHeartbeatV2) SupportsFallback() bool {
+	return h.announcer != nil
+}
+
+func (h *appServerHeartbeatV2) FallbackAnnounce(ctx context.Context) (ok bool) {
+	if h.announcer == nil {
+		return false
+	}
+	server := h.getServer(ctx)
+	_, err := h.announcer.UpsertApplicationServer(ctx, server)
+	if err != nil {
+		if !errors.Is(err, context.Canceled) && status.Code(err) != codes.Canceled {
+			log.Warnf("Failed to perform fallback heartbeat for app server: %v", err)
+		}
+		return false
+	}
+	h.prev = server
+	return true
+}
+
+var (
+	minAppVersion15 = semver.New("15.3.0")
+	minAppVersion14 = semver.New("14.3.17")
+	minAppVersion13 = semver.New("13.4.23")
+)
+
+func (h *appServerHeartbeatV2) Announce(ctx context.Context, sender inventory.DownstreamSender) (ok bool) {
+	authVersion, err := semver.NewVersion(sender.Hello().Version)
+	if err != nil {
+		return false
+	}
+
+	// AppServer heartbeats via inventory control stream were not introduced in a major version,
+	// so there is a chance that the Auth server is unable to process the request via the inventory
+	// control stream. If the Auth server is detected to be running an incompatible version, then use
+	// the fallback mechanism.
+	// TODO(tross) DELETE IN 16.0.0
+	if authVersion.LessThan(*minAppVersion15) || authVersion.LessThan(*minAppVersion14) || authVersion.LessThan(*minAppVersion13) {
```

Comment:
I think this will always be true for anything less than the v15 cutoff, right?  I believe that to achieve what you want here we should be switching on the auth server's major version first.